File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/playlist/PlayListPresenter.java
package com.github.anrimian.musicplayer.ui.playlist_screens.playlist;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.github.anrimian.musicplayer.domain.business.player.MusicPlayerInteractor;
import com.github.anrimian.musicplayer.domain.business.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.business.settings.DisplaySettingsInteractor;
import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.composition.PlayQueueEvent;
import com.github.anrimian.musicplayer.domain.models.composition.PlayQueueItem;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem;
import com.github.anrimian.musicplayer.domain.utils.model.Item;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import javax.annotation.Nullable;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

import static com.github.anrimian.musicplayer.data.utils.rx.RxUtils.isInactive;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.asList;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.mapList;
import static java.util.Objects.requireNonNull;

@InjectViewState
public class PlayListPresenter extends MvpPresenter<PlayListView> {

    private final MusicPlayerInteractor playerInteractor;
    private final PlayListsInteractor playListsInteractor;
    private final DisplaySettingsInteractor displaySettingsInteractor;
    private final ErrorParser errorParser;
    private final Scheduler uiScheduler;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();
    private final CompositeDisposable presenterBatterySafeDisposable = new CompositeDisposable();
    private Disposable currentItemDisposable;

    private final long playListId;

    private List<PlayListItem> items = new ArrayList<>();

    @Nullable
    private PlayList playList;

    private final List<Composition> compositionsForPlayList = new LinkedList<>();
    private final List<Composition> compositionsToDelete = new LinkedList<>();

    private int startDragPosition;

    private PlayQueueItem currentItem;

    private Item<PlayListItem> deletedItem;

    public PlayListPresenter(long playListId,
                             MusicPlayerInteractor playerInteractor,
                             PlayListsInteractor playListsInteractor,
                             DisplaySettingsInteractor displaySettingsInteractor,
                             ErrorParser errorParser,
                             Scheduler uiScheduler) {
        this.playListId = playListId;
        this.playerInteractor = playerInteractor;
        this.playListsInteractor = playListsInteractor;
        this.displaySettingsInteractor = displaySettingsInteractor;
        this.errorParser = errorParser;
        this.uiScheduler = uiScheduler;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        subscribeOnCompositions();
        subscribePlayList();
        subscribeOnCurrentComposition();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onStart() {
        if (!items.isEmpty()) {
            subscribeOnCurrentComposition();
        }
    }

    void onStop() {
        presenterBatterySafeDisposable.clear();
    }

    void onCompositionClicked(PlayListItem playListItem, int position) {
        getViewState().showCompositionActionDialog(playListItem, position);
    }

    void onItemIconClicked(int position) {
        playerInteractor.startPlaying(mapList(items, PlayListItem::getComposition), position);
    }

    void onPlayAllButtonClicked() {
        playerInteractor.startPlaying(mapList(items, PlayListItem::getComposition));
    }

    void onDeleteCompositionButtonClicked(Composition composition) {
        compositionsToDelete.clear();
        compositionsToDelete.add(composition);
        getViewState().showConfirmDeleteDialog(compositionsToDelete);
    }

    void onDeleteCompositionsDialogConfirmed() {
        deletePreparedCompositions();
    }

    void onAddToPlayListButtonClicked(Composition composition) {
        compositionsForPlayList.clear();
        compositionsForPlayList.add(composition);
        getViewState().showSelectPlayListDialog();
    }

    void onPlayNextCompositionClicked(Composition composition) {
        addCompositionsToPlayNext(asList(composition));
    }

    void onAddToQueueCompositionClicked(Composition composition) {
        addCompositionsToEnd(asList(composition));
    }

    void onPlayListToAddingSelected(PlayList playList) {
        addPreparedCompositionsToPlayList(playList);
    }

    void onDeleteFromPlayListButtonClicked(PlayListItem playListItem, int position) {
        deleteItem(playListItem, position);
    }

    void onDeletePlayListButtonClicked() {
        getViewState().showConfirmDeletePlayListDialog(playList);
    }

    void onDeletePlayListDialogConfirmed() {
        playListsInteractor.deletePlayList(requireNonNull(playList).getId())
                .observeOn(uiScheduler)
                .subscribe(this::onPlayListDeleted, this::onPlayListDeletingError);
    }

    void onFragmentMovedToTop() {
        playListsInteractor.setSelectedPlayListScreen(playListId);
    }

    void onItemSwipedToDelete(int position) {
        deleteItem(items.get(position), position);
    }

    void onItemMoved(int from, int to) {
        if (from < to) {
            for (int i = from; i < to; i++) {
                swapItems(i, i + 1);
            }
        } else {
            for (int i = from; i > to; i--) {
                swapItems(i, i - 1);
            }
        }
    }

    void onDragStarted(int position) {
        startDragPosition = position;
    }

    void onDragEnded(int position) {
        playListsInteractor.moveItemInPlayList(playList, startDragPosition, position);
    }

    void onPlayActionSelected(int position) {
        playerInteractor.startPlaying(
                mapList(items, PlayListItem::getComposition),
                position
        );
    }

    void onRestoreRemovedItemClicked() {
        Composition removedComposition = deletedItem.getData().getComposition();
        playListsInteractor.addCompositionsToPlayList(asList(removedComposition),
                playList,
                deletedItem.getPosition())
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
    }

    void onChangePlayListNameButtonClicked() {
        if (playList != null) {
            getViewState().showEditPlayListNameDialog(playList);
        }
    }

    boolean isCoversEnabled() {
        return displaySettingsInteractor.isCoversEnabled();
    }

    private void addCompositionsToPlayNext(List<Composition> compositions) {
        playerInteractor.addCompositionsToPlayNext(compositions)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
    }

    private void addCompositionsToEnd(List<Composition> compositions) {
        playerInteractor.addCompositionsToEnd(compositions)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
    }

    private void onDefaultError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showErrorMessage(errorCommand);
    }

    private void deleteItem(PlayListItem playListItem, int position) {
        playListsInteractor.deleteItemFromPlayList(playListItem, playListId)
                .observeOn(uiScheduler)
                .subscribe(() -> onDeleteItemCompleted(playListItem, position), this::onDeleteItemError);
    }

    private void swapItems(int from, int to) {
        Collections.swap(items, from, to);
        getViewState().notifyItemMoved(from, to);
    }

    private void onPlayListDeletingError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showDeletePlayListError(errorCommand);
    }

    private void onPlayListDeleted() {
        getViewState().showPlayListDeleteSuccess(playList);
    }

    private void onDeleteItemCompleted(PlayListItem item, int position) {
        if (playList != null) {
            deletedItem = new Item<>(item, position);
//            getViewState().notifyItemRemoved(position);
            getViewState().showDeleteItemCompleted(playList, asList(item));
        }
    }

    private void onDeleteItemError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showDeleteItemError(errorCommand);
    }

    private void addPreparedCompositionsToPlayList(PlayList playList) {
        playListsInteractor.addCompositionsToPlayList(compositionsForPlayList, playList)
                .observeOn(uiScheduler)
                .subscribe(() -> onAddingToPlayListCompleted(playList),
                        this::onAddingToPlayListError);
    }

    private void onAddingToPlayListError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showAddingToPlayListError(errorCommand);
    }

    private void onAddingToPlayListCompleted(PlayList playList) {
        getViewState().showAddingToPlayListComplete(playList, compositionsForPlayList);
        compositionsForPlayList.clear();
    }

    private void deletePreparedCompositions() {
        playerInteractor.deleteCompositions(compositionsToDelete)
                .observeOn(uiScheduler)
                .subscribe(this::onDeleteCompositionsSuccess, this::onDeleteCompositionsError);
    }

    private void onDeleteCompositionsSuccess() {
        getViewState().showDeletedCompositionMessage(compositionsToDelete);
    }

    private void onDeleteCompositionsError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showDeleteCompositionError(errorCommand);
    }

    private void subscribeOnCompositions() {
        getViewState().showLoading();
        presenterDisposable.add(playListsInteractor.getCompositionsObservable(playListId)
                .observeOn(uiScheduler)
                .subscribe(this::onPlayListsReceived,
                        t -> getViewState().closeScreen(),
                        getViewState()::closeScreen));
    }

    private void subscribePlayList() {
        presenterDisposable.add(playListsInteractor.getPlayListObservable(playListId)
                .observeOn(uiScheduler)
                .subscribe(this::onPlayListInfoReceived,
                        t -> getViewState().closeScreen(),
                        getViewState()::closeScreen));
    }

    private void onPlayListInfoReceived(PlayList playList) {
        this.playList = playList;
        getViewState().showPlayListInfo(playList);
    }

    private void onPlayListsReceived(List<PlayListItem> list) {
        this.items = list;
        getViewState().updateItemsList(list);
        if (items.isEmpty()) {
            getViewState().showEmptyList();
        } else {
            getViewState().showList();

            if (isInactive(currentItemDisposable)) {
                subscribeOnCurrentComposition();
            }
        }
    }

    private void subscribeOnCurrentComposition() {
        currentItemDisposable = playerInteractor.getCurrentCompositionObservable()
                .observeOn(uiScheduler)
                .subscribe(this::onCurrentCompositionReceived, errorParser::logError);
        presenterBatterySafeDisposable.add(currentItemDisposable);
    }

    private void onCurrentCompositionReceived(PlayQueueEvent playQueueEvent) {
        currentItem = playQueueEvent.getPlayQueueItem();
    }

}

File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/playlist/PlayListPresenter.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.playlist

import com.github.anrimian.musicplayer.data.utils.rx.RxUtils
import com.github.anrimian.musicplayer.domain.interactors.player.MusicPlayerInteractor
import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.interactors.settings.DisplaySettingsInteractor
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueEvent
import com.github.anrimian.musicplayer.domain.models.play_queue.PlayQueueItem
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem
import com.github.anrimian.musicplayer.domain.utils.model.Item
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import moxy.InjectViewState
import java.util.*

@InjectViewState
class PlayListPresenter(private val playListId: Long,
                        private val playerInteractor: MusicPlayerInteractor,
                        private val playListsInteractor: PlayListsInteractor,
                        private val displaySettingsInteractor: DisplaySettingsInteractor,
                        errorParser: ErrorParser,
                        uiScheduler: Scheduler)
    : AppPresenter<PlayListView>(uiScheduler, errorParser) {

    private val presenterBatterySafeDisposable = CompositeDisposable()

    private var currentItemDisposable: Disposable? = null
    private var items: List<PlayListItem> = ArrayList()
    private var playList: PlayList? = null

    private val compositionsForPlayList: MutableList<Composition> = LinkedList()
    private val compositionsToDelete: MutableList<Composition> = LinkedList()

    private var startDragPosition = 0
    private var currentItem: PlayQueueItem? = null
    private var deletedItem: Item<PlayListItem>? = null

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        subscribeOnCompositions()
        subscribePlayList()
        subscribeOnCurrentComposition()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenterDisposable.dispose()
    }

    fun onStart() {
        if (items.isNotEmpty()) {
            subscribeOnCurrentComposition()
        }
    }

    fun onStop() {
        presenterBatterySafeDisposable.clear()
    }

    fun onCompositionClicked(playListItem: PlayListItem?, position: Int) {
        viewState.showCompositionActionDialog(playListItem, position)
    }

    fun onItemIconClicked(position: Int) {
        playerInteractor.startPlaying(items.map(PlayListItem::getComposition), position)
    }

    fun onPlayAllButtonClicked() {
        playerInteractor.startPlaying(items.map(PlayListItem::getComposition))
    }

    fun onDeleteCompositionButtonClicked(composition: Composition) {
        compositionsToDelete.clear()
        compositionsToDelete.add(composition)
        viewState.showConfirmDeleteDialog(compositionsToDelete)
    }

    fun onDeleteCompositionsDialogConfirmed() {
        deletePreparedCompositions()
    }

    fun onAddToPlayListButtonClicked(composition: Composition) {
        compositionsForPlayList.clear()
        compositionsForPlayList.add(composition)
        viewState.showSelectPlayListDialog()
    }

    fun onPlayNextCompositionClicked(composition: Composition) {
        addCompositionsToPlayNext(listOf(composition))
    }

    fun onAddToQueueCompositionClicked(composition: Composition) {
        addCompositionsToEnd(listOf(composition))
    }

    fun onPlayListToAddingSelected(playList: PlayList) {
        addPreparedCompositionsToPlayList(playList)
    }

    fun onDeleteFromPlayListButtonClicked(playListItem: PlayListItem, position: Int) {
        deleteItem(playListItem, position)
    }

    fun onDeletePlayListButtonClicked() {
        viewState.showConfirmDeletePlayListDialog(playList)
    }

    fun onDeletePlayListDialogConfirmed() {
        playListsInteractor.deletePlayList(playList!!.id)
                .subscribeOnUi(this::onPlayListDeleted, this::onPlayListDeletingError)
    }

    fun onFragmentMovedToTop() {
        playListsInteractor.setSelectedPlayListScreen(playListId)
    }

    fun onItemSwipedToDelete(position: Int) {
        deleteItem(items[position], position)
    }

    fun onItemMoved(from: Int, to: Int) {
        if (from < to) {
            for (i in from until to) {
                swapItems(i, i + 1)
            }
        } else {
            for (i in from downTo to + 1) {
                swapItems(i, i - 1)
            }
        }
    }

    fun onDragStarted(position: Int) {
        startDragPosition = position
    }

    fun onDragEnded(position: Int) {
        playListsInteractor.moveItemInPlayList(playList, startDragPosition, position)
    }

    fun onPlayActionSelected(position: Int) {
        playerInteractor.startPlaying(items.map(PlayListItem::getComposition), position)
    }

    fun onRestoreRemovedItemClicked() {
        val removedComposition = deletedItem!!.data!!.composition
        playListsInteractor.addCompositionsToPlayList(listOf(removedComposition),
                        playList,
                        deletedItem!!.position)
                .justSubscribe(this::onDefaultError)
    }

    fun onChangePlayListNameButtonClicked() {
        if (playList != null) {
            viewState.showEditPlayListNameDialog(playList)
        }
    }

    fun isCoversEnabled() = displaySettingsInteractor.isCoversEnabled

    private fun addCompositionsToPlayNext(compositions: List<Composition>) {
        playerInteractor.addCompositionsToPlayNext(compositions)
                .subscribeOnUi(viewState::onCompositionsAddedToPlayNext, this::onDefaultError)
    }

    private fun addCompositionsToEnd(compositions: List<Composition>) {
        playerInteractor.addCompositionsToEnd(compositions)
                .subscribeOnUi(viewState::onCompositionsAddedToPlayNext, this::onDefaultError)
    }

    private fun onDefaultError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showErrorMessage(errorCommand)
    }

    private fun deleteItem(playListItem: PlayListItem, position: Int) {
        playListsInteractor.deleteItemFromPlayList(playListItem, playListId)
                .subscribeOnUi({ onDeleteItemCompleted(playListItem, position) }, this::onDeleteItemError)
    }

    private fun swapItems(from: Int, to: Int) {
        Collections.swap(items, from, to)
        viewState.notifyItemMoved(from, to)
    }

    private fun onPlayListDeletingError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showDeletePlayListError(errorCommand)
    }

    private fun onPlayListDeleted() {
        viewState.showPlayListDeleteSuccess(playList)
    }

    private fun onDeleteItemCompleted(item: PlayListItem, position: Int) {
        if (playList != null) {
            deletedItem = Item(item, position)
            viewState.showDeleteItemCompleted(playList, listOf(item))
        }
    }

    private fun onDeleteItemError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showDeleteItemError(errorCommand)
    }

    private fun addPreparedCompositionsToPlayList(playList: PlayList) {
        playListsInteractor.addCompositionsToPlayList(compositionsForPlayList, playList)
                .subscribeOnUi({ onAddingToPlayListCompleted(playList) }, this::onAddingToPlayListError)
    }

    private fun onAddingToPlayListError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showAddingToPlayListError(errorCommand)
    }

    private fun onAddingToPlayListCompleted(playList: PlayList) {
        viewState.showAddingToPlayListComplete(playList, compositionsForPlayList)
        compositionsForPlayList.clear()
    }

    private fun deletePreparedCompositions() {
        playerInteractor.deleteCompositions(compositionsToDelete)
                .subscribeOnUi(this::onDeleteCompositionsSuccess, this::onDeleteCompositionsError)
    }

    private fun onDeleteCompositionsSuccess() {
        viewState.showDeletedCompositionMessage(compositionsToDelete)
    }

    private fun onDeleteCompositionsError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showDeleteCompositionError(errorCommand)
    }

    private fun subscribeOnCompositions() {
        viewState.showLoading()
        playListsInteractor.getCompositionsObservable(playListId)
                .subscribeOnUi(
                        this::onPlayListsReceived,
                        { viewState.closeScreen() },
                        viewState::closeScreen
                )
    }

    private fun subscribePlayList() {
        playListsInteractor.getPlayListObservable(playListId)
                .subscribeOnUi(
                        this::onPlayListInfoReceived,
                        { viewState.closeScreen() },
                        viewState::closeScreen
                )
    }

    private fun onPlayListInfoReceived(playList: PlayList) {
        this.playList = playList
        viewState.showPlayListInfo(playList)
    }

    private fun onPlayListsReceived(list: List<PlayListItem>) {
        items = list
        viewState.updateItemsList(list)
        if (items.isEmpty()) {
            viewState.showEmptyList()
        } else {
            viewState.showList()
            if (RxUtils.isInactive(currentItemDisposable)) {
                subscribeOnCurrentComposition()
            }
        }
    }

    private fun subscribeOnCurrentComposition() {
        currentItemDisposable = playerInteractor.currentQueueItemObservable
                .observeOn(uiScheduler)
                .subscribe(this::onCurrentCompositionReceived, errorParser::logError)
        presenterBatterySafeDisposable.add(currentItemDisposable!!)
    }

    private fun onCurrentCompositionReceived(playQueueEvent: PlayQueueEvent) {
        currentItem = playQueueEvent.playQueueItem
    }

}